core/templates/rid_owner.h

```diff
@@ -43,6 +43,39 @@
 #include <stdio.h>
 #include <typeinfo>
 
+#ifdef SANITIZERS_ENABLED
+#ifdef __has_feature
+#if __has_feature(thread_sanitizer)
+#define TSAN_ENABLED
+#endif
+#elif defined(__SANITIZE_THREAD__)
+#define TSAN_ENABLED
+#endif
+#endif
+
+#ifdef TSAN_ENABLED
+#include <sanitizer/tsan_interface.h>
+#endif
+
+// The following macros would need to be implemented somehow
+// for purely weakly ordered architectures. There's a test case
+// ("[RID_Owner] Thread safety") with potential to catch issues
+// on such architectures if these primitives fail to be implemented.
+// For now, they will be just markers about needs that may arise.
+#define WEAK_MEMORY_ORDER 0
+#if WEAK_MEMORY_ORDER
+// Ideally, we'd have implementations that collaborate with the
+// sync mechanism used (e.g., the mutex) so instead of some full
+// memory barriers being issued, some acquire-release on the
+// primitive itself. However, these implementations will at least
+// provide correctness.
+#define SYNC_ACQUIRE std::atomic_thread_fence(std::memory_order_acquire);
+#define SYNC_RELEASE std::atomic_thread_fence(std::memory_order_release);
+#else
+#define SYNC_ACQUIRE
+#define SYNC_RELEASE
```

Comment:
This is not enough on systems with strong memory models.

Not because of HW reasons, but compiler ones. x86 will not reorder your memory operations, but compiler optimizations still can.

You still need `std::atomic_thread_fence()` to tell the compiler not to reorder stuff in ways that would break your code.

Reply:
Since the acquire and release points are at the beginning and end of functions, I was just relying on functions calls acting as compiler barriers. That said, I don't think some extra safety will cause any harm. I've updated the code. However, I'm adding compiler barriers there (`std::atomic_signal_fence()`) instead.